### ISSUE TITLE
Add defer lock: keep some tasks for being enqueued more than once

### ISSUE THREAD

**Issue by elemoine** [peopledoc/procrastinate#217]
(Edit @ewjoachim: Original title was "Introduce queue size?")

Here's my use-case:

Worker healthcheck tasks, launched periodically (every 2 minutes) by a cron job, through a dedicated Procrastinate queue. If the worker is down, healthcheck jobs will accumulate in the queue. When the Procrastinate worker is up again, it will execute all the healthcheck tasks in the queue, resulting in a burst of healthchecks. Instead, I'd like the worker to execute only one healthcheck task.

For this problem, I've been contemplating the idea of introducing the notion of queue size. On a defer, if the queue size is exceeded, the defer operation will basically be a noop, meaning that no new job will be inserted into the procrastinate_jobs table.

And this is the "defer_job" SQL query I've been considering for the case when a queue size is defined on the defer operation:

```sql
INSERT INTO procrastinate_jobs (queue_name, task_name, lock, args, scheduled_at)
SELECT %(queue)s, %(task_name)s, %(lock)s, %(args)s, %(scheduled_at)s
WHERE (
    SELECT COUNT(*)
        FROM procrastinate_jobs
       WHERE status = 'todo' AND queue_name = %(queue)s
) < %(queue_size)s RETURNING id;
```

I think there's a possible race condition, when multiple defer operations occur at the time. Simultaneous defer operations may see a queue that is not full, and therefore insert a new job into the queue, possibly exceeding the queue size. But that might not be a problem for the use-cases we want to support with this new functionality.

Interested in discussing this use-case, and the "queue size" functionality I've been considering for it.

**Comment by elemoine** [comment]
@ewjoachim and I discussed this, and we want to approach this in a different way. The plan is to introduce a new text column named "defer_lock" (or some other name) to the procrastinate_jobs table, and a constraint preventing from having multiple jobs in the "todo" state with the same value for that new column.

For example, assuming we have a "task_a". "task_a" is deferred with the value "a" for "defer_lock". Deferring "task_a" again will raise an exception until the first "task_a" job transitions from "todo" to "doing".

**Comment by ewjoachim** [comment]
Yes. Adding 2 details: we were wondering whether there should be an error if we deferred a task with the log taken. Considering the zen of Python

> Errors should never pass silently, unless explicitly silenced

We decided that you always know that you're deferring a task with a defer lock so it's always easy to expect an exception there, and then you get to decide if you want to ignore it or react. This is always something we could revisit later.

Also, we may implement limit on queue sizes but rather for performance reasons. This will be another unrelated ticket.
